pallets/parachain-staking/src/types.rs

```diff
@@ -0,0 +1,1673 @@
+            delegations: Vec::new(),
+            total: B::default(),
+        }
+    }
+}
+
+impl<AccountId, Balance: Copy + Ord + sp_std::ops::AddAssign + Zero + Saturating>
+    Delegations<AccountId, Balance>
+{
+    pub fn sort_greatest_to_least(&mut self) {
+        self.delegations.sort_by(|a, b| b.amount.cmp(&a.amount));
+    }
+    /// Insert sorted greatest to least and increase .total accordingly
+    /// Insertion respects first come first serve so new delegations are pushed after existing
+    /// delegations if the amount is the same
+    pub fn insert_sorted_greatest_to_least(&mut self, delegation: Bond<AccountId, Balance>) {
+        self.total = self.total.saturating_add(delegation.amount);
+        // if delegations nonempty && last_element == delegation.amount => push input and return
+        if !self.delegations.is_empty() {
+            // if last_element == delegation.amount => push the delegation and return early
+            if self.delegations[self.delegations.len() - 1].amount == delegation.amount {
+                self.delegations.push(delegation);
+                // early return
+                return;
+            }
+        }
+        // else binary search insertion
+        match self
+            .delegations
+            .binary_search_by(|x| delegation.amount.cmp(&x.amount))
+        {
+            // sorted insertion on sorted vec
+            // enforces first come first serve for equal bond amounts
+            Ok(i) => {
+                let mut new_index = i + 1;
+                while new_index <= (self.delegations.len() - 1) {
+                    if self.delegations[new_index].amount == delegation.amount {
+                        new_index = new_index.saturating_add(1);
+                    } else {
+                        self.delegations.insert(new_index, delegation);
+                        return;
+                    }
+                }
+                self.delegations.push(delegation)
+            }
+            Err(i) => self.delegations.insert(i, delegation),
+        }
+    }
+    /// Return the capacity status for top delegations
+    pub fn top_capacity<T: Config>(&self) -> CapacityStatus {
+        match &self.delegations {
+            x if x.len() as u32 >= T::MaxTopDelegationsPerCandidate::get() => CapacityStatus::Full,
+            x if x.is_empty() => CapacityStatus::Empty,
+            _ => CapacityStatus::Partial,
+        }
+    }
+    /// Return the capacity status for bottom delegations
+    pub fn bottom_capacity<T: Config>(&self) -> CapacityStatus {
+        match &self.delegations {
+            x if x.len() as u32 >= T::MaxBottomDelegationsPerCandidate::get() => {
+                CapacityStatus::Full
+            }
+            x if x.is_empty() => CapacityStatus::Empty,
+            _ => CapacityStatus::Partial,
+        }
+    }
+    /// Return last delegation amount without popping the delegation
+    pub fn lowest_delegation_amount(&self) -> Balance {
```

Comment:
If delegations.is_empty, it still returns zero.
But if `delegations` is never empty, ignore this comment.